Tasks/JavaToolInstaller/javatoolinstaller.ts

```diff
@@ -0,0 +1,101 @@
+taskLib.setResourcePath(path.join(__dirname, 'task.json'));
+
+async function run() {
+    try {
+        let versionSpec = taskLib.getInput('versionSpec', true);
+        await getJava(versionSpec);
+    }
+    catch (error) {
+        console.error('ERR:' + error.message);
+    } 
+}
+
+async function getJava(versionSpec: string) {
+    toolLib.debug('Trying to get tool from local cache');
+    let localVersions: string[] = toolLib.findLocalToolVersions('Java');
+    let version: string = toolLib.evaluateVersions(localVersions, versionSpec);
+    let fromAzure: boolean = ('AzureStorage' == taskLib.getInput('jdkSource', true));
+    let fromLocalDirectory: boolean = ('LocalDirectory' == taskLib.getInput('jdkSource', true));
+    let extractLocation: string = taskLib.getPathInput('destinationFolder', true);
+    let fileName: string;
+    let cleanDestinationFolder: boolean = taskLib.getBoolInput('cleanDestinationFolder', false);
+    let fileEnding: string;
+
+     // Clean the destination folder before downloading and extracting?
+     if (cleanDestinationFolder && taskLib.exist(extractLocation)) {
+        console.log(taskLib.loc('CleanDestDir', extractLocation));
+        taskLib.rmRF(extractLocation);
+    }
+
+    if (version) { //This version of Java JDK is already in the cache. Use it instead of downloading again.
+        console.log(taskLib.loc("Info_ResolvedToolFromCache", version));
+    }
+    else if (fromAzure) { //Download JDK from an Azure blob storage location and extract.
+        try {
+            fileEnding = (taskLib.getInput('fileType', true) == 'compressedTar') ? ".tar.gz" : ("." + taskLib.getInput('fileType', true));
+            
+            await new AzureStorageArtifactDownloader().downloadArtifacts(extractLocation, fileEnding);
+
+            var extractSource = buildFilePath(extractLocation, fileEnding);
+            new JavaFilesExtractor().unzipJavaDownload(extractSource, fileEnding);
+            fileName = taskLib.getInput('commonVirtualPath', true).split(/[\\\/]/).pop();
+        } catch (err) {
+            taskLib.debug(err.message);
+            taskLib.error(err.message);
+            taskLib.setResult(taskLib.TaskResult.Failed, err.message);
+        }
+    }
+    else if (fromLocalDirectory) { //JDK is in a local directory. Extract to specified target directory.
+        fileName = taskLib.getInput('jdkPath', true).split(/[\\\/]/).pop();
+        fileEnding = getFileEnding(fileName);
+        new JavaFilesExtractor().unzipJavaDownload(taskLib.getInput('jdkPath', true), fileEnding);
+        fileName = fileName.replace(fileEnding, '');
+        getJavaHomePath(taskLib.getInput('jdkPath', true));
+    }
+
+    var filePath = path.normalize(taskLib.getPathInput('destinationFolder', true, false).trim());
+    var toolPath = path.join(filePath, fileName);
+    console.log("JAVA_HOME is being set to: " + toolPath);
+    taskLib.setVariable('JAVA_HOME', toolPath);
```

Comment:
Additionally set JAVA_HOME_<JDK_version>_<JDK_Architecture> to enable other java tasks to find it: https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/Common/java-common/java-common.ts#L39